### PR TITLE
Use 128 bits for TypeId hash

### PR DESCRIPTION
**Pull request by thomcc**
Preliminary/Draft impl of https://github.com/rust-lang/compiler-team/issues/608

Prior art (probably incomplete list)
- https://github.com/rust-lang/rust/pull/75923
- https://github.com/rust-lang/rust/pull/95845

### CODE REVIEW

**Comment by thomcc** [comment]
There are a couple test failures that hit locally. Just two, really: https://gist.github.com/thomcc/514a6b46f9636af96cf0c5f9fdd0f37f. Not sure about them.

Other things I wasn't sure about:
1. Why do I need to handle the intrinsic in `rustc_codegen_llvm` (I thought that it should be lowered already, since I did so in `rustc_codegen_ssa`).
2. If there's some way to avoid adding `type_id128` as an intrinsic, and instead just change the return type of `type_id`. I couldn't see a clean way to make that work in the compiler with bootstrapping, but maybe I didn't try hard enough or just confused myself with it.

**Comment by thomcc** [comment]
Oh, it fails because tracing-subscriber relies on TypeId being u64. :|

**Comment by thomcc** [comment]
I've added a hacky `impl Hash for TypeId` which should avoid upsetting `tracing-subscriber`.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs
Comment (compiler-errors) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: I'm curious to know what fails when this is missing?

Comment (scottmcm) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: Definitely very surprising.  The things in codegen_ssa are attempted first, which you've updated, so I agree this shouldn't be needed.

File: compiler/rustc_middle/src/ty/util.rs
Comment (scottmcm) [inline compiler/rustc_middle/src/ty/util.rs:142]: YMMV: maybe make this generic, returning `<W: StableHasherResult>`, to avoid needing to copy-paste the implementation?

File: compiler/rustc_codegen_llvm/src/intrinsic.rs
Comment (thomcc) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: `error: internal compiler error: compiler/rustc_codegen_llvm/src/intrinsic.rs:381:18: unknown intrinsic 'type_id128' -- should it have been lowered earlier?`

Comment (compiler-errors) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: Sorry, I mean when exactly it's failing. Stage 0 library build, stage 1 library build (core, std, later?), or a specific ui test?

Also, the whole stack trace may be useful too.

Comment (thomcc) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: Oh, during `stage1` build of library (specifically `alloc`): https://gist.github.com/thomcc/1408507eab7b6e0b82ab97670236663a

Comment (compiler-errors) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:163]: Huh -- then @scottmcm is right. We should be handling this at compiler/rustc_codegen_ssa/src/mir/intrinsic.rs:130...

Comment (thomcc) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs]: Hm, it might not be needed. Things seem to pass locally without it now, I wonder if I had forgotten to save a file somewhere?

Comment (scottmcm) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs]: I just pulled down the changes locally and it built+tested just fine for me too.

File: compiler/rustc_middle/src/ty/util.rs
Comment (thomcc) [reply compiler/rustc_middle/src/ty/util.rs]: I realized that I'm being dumb and don't need both `type_id` and `type_id128`, so this is a moot point.

File: library/core/src/any.rs
Comment (eddyb) [inline library/core/src/any.rs:708]: This might work, but IIRC in #95845 (or some other PR), I made it hash just 64 bits.
One thing to note is that this being *a hash*, more bits isn't more better, and you can avoid any perf complaints by only hashing 64 bits just like before.

IOW, it doesn't have to be *perfect*, as `Eq` will still compare the full width, you just need to feed at least 64 bits to the hasher ([`Hasher::finish`](https://doc.rust-lang.org/std/hash/trait.Hasher.html#tymethod.finish) returns `u64` so I suspect that more bits than that is overkill).

Comment (thomcc) [reply library/core/src/any.rs:708]: > This might work, but IIRC in https://github.com/rust-lang/rust/pull/95845 (or some other PR), I made it hash just 64 bits.

Hm, that's reasonable. I feel like it's better to hash the full value, but I don't feel that strongly about it.

Comment (thomcc) [reply library/core/src/any.rs:708]: > [`Hasher::finish`](https://doc.rust-lang.org/std/hash/trait.Hasher.html#tymethod.finish) returns `u64` so I suspect that more bits than that is overkill.

This is a really compelling point, honestly.

Comment (thomcc) [reply library/core/src/any.rs]: Compelling enough that I changed it. I still think the impl should get t-libs sign-off though, but not worth doing that prior to the MCP being accepted.

**Comment by thomcc** [comment]
Going to do a perf run, since it was mentioned as a concern.

@bors try @rust-timer queue

File: library/core/src/any.rs
Comment (thomcc) [inline library/core/src/any.rs:722]: Don't land without me doing this (but it probably would be premature before the MCP completes).

**Comment by bjorn3** [comment]
cc @alice-i-cecile maybe this should be benchmarked with Bevy?

**Comment by eddyb** [comment]
cc-ing @Badel2 @CAD97 @m-ou-se @digama0 @RalfJung @nikomatsakis @joshtriplett (based on comments you each had on #95845 regarding hashes - apologies if this PR isn't relevant, but I wanted to at least connect those dots)

**Review by davidtwco — APPROVED** [review]
Implementation LGTM

**Comment by Badel2** [comment]
Wow, so excited to see this finally being fixed!

Regarding hashing only the lower u64 in the Hash impl, that it is a common and perfectly valid way to truncate a cryptographically secure hash function, it is even in the standard for SHA-2. And even if the hash function used by TypeId is not cryptographically secure, the worst attack which is creating hash collisions will have the same cost as if we were hashing the full u128 since the output of Hash::finish is a u64 anyway.

The only problem is that it is possible to create a TypeId using transmute, and in that case creating a collision is free because we can modify the output of the first hash function. (transmute TypeId to u128, modify without touching lower u64, transmute back to TypeId). The impact of that problem is that you can make a hashmap O(n^2), but only in your own code, so I will let someone else decide if it is worth to fix it.

After taking a look at the offending code that breaks, here: https://github.com/tokio-rs/tracing/blob/3392be93cb9262a5da54c23e5772e79a8736d50e/tracing-subscriber/src/registry/extensions.rs#L27

I propose to just call write_u64 twice. This will avoid worrying about the hash collisions created using transmute, and it will not break tracing-subscriber, which will only use the input of the second call as the hash.

Also the last time I remember talking about this, the problem was that some crates are using transmute between TypeId and u64, how is that fixed here? Was that fixed in those crates at some point?

Thank you for moving this issue forward!

**Comment by thomcc** [comment]
> I propose to just call write_u64 twice. This will avoid worrying about the hash collisions created using transmute, and it will not break tracing-subscriber, which will only use the input of the second call as the hash.

This is what I did initially, after looking at the same code. That said, I find @eddyb's argument that it's fine to just use 64 bits pretty compelling.

> the worst attack which is creating hash collisions will have the same cost as if we were hashing the full u128 since the output of Hash::finish is a u64 anyway

We're not really concerned about attacks, so much as accidental collisions.

> the problem was that some crates are using transmute between TypeId and u64, how is that fixed here

It's not. Those crates are going to be broken no matter how we fix this issue (even the string pointer approach probably won't work right for them, although I suppose on 64 bit targets they might be less broken), the only way to keep them working is to decide this never changes, which seems unreasonable.

In the past we haven't really considered that reason to avoid a motivated fix; for example, plenty of code used to do type punning between std::net types and libc... although it's possible the breakage in this case could be too much to bear.

It's worth doing a crater run for this at least, if for no other reason than to give such code a heads up.

**Comment by thomcc** [comment]
Starting a try for a crater run to determine the impact of this.

@bors try

**Comment by bjorn3** [comment]
> > the worst attack which is creating hash collisions will have the same cost as if we were hashing the full u128 since the output of Hash::finish is a u64 anyway
> 
> We're not really concerned about attacks, so much as accidental collisions.

For `Eq` collisions are unsound. For `Hash` collisions at most you did see a (severe) performance degradation. So for as long as `Eq` uses the full 128bit hash, `Hash` can reduce it to 64bit without all that much issues.

**Comment by eddyb** [comment]
Regarding crater, note that https://github.com/rust-lang/rust/pull/95845#issuecomment-1117219883 said:
> That crater run showed that the breakage is small, and almost all crates that were broken have either been fixed or only use it in a test, meaning it won't break crates that depend on them.

And it's in relation to this crater run: https://github.com/rust-lang/rust/pull/75923#issuecomment-699064557
(also, that older crater'd PR was "Widen TypeId from 64 bits to 128." i.e. same as this PR)

**Comment by thomcc** [comment]
> For Eq collisions are unsound. For Hash collisions at most you did see a (severe) performance degradation. So for as long as Eq uses the full 128bit hash, Hash can reduce it to 64bit without all that much issues.

Right, I state as much in the comment in the code. That was also intended replying to the comments they made in the next paragraph about malicious actors forging TypeId for other reasons. There's no reason for us to care about that.

**Comment by CAD97** [comment]
Since pinged, I see no reason to object to the widening of the `TypeId` hash. (Although I do find it interesting that the impl didn't change, other than to adjust the type; presumably then the compiler's stable hash API is sufficiently generic over return type, unlike the `Hash` trait which always produces `u64`.) If there's any reason to believe the lower or higher `u64` part of the `u128` hash to be higher quality, the higher quality part should be used. If they're believed to be equally uniformly distributed, truncating the hash either way is plainly acceptable.

I do think it would be interesting to investigate a "perfect" solution[^x] in the future, but the potential of doing so absolutely shouldn't block this improvement.

[^x]: i.e. using the target platform's choice for implementing C++ RTTI, assuming that exists, e.g. (ab)using symbol deduplication or comparing mangled type names. Although that should probably be limited to a codegen option, since developers may want to ensure type names are all stripped and I'm fairly certain symbol deduplication schemes don't play well with runtime loading of dynamic libraries.

**Comment by alice-i-cecile** [comment]
I've requested a benchmark run of this for Bevy @bjorn3. I'm not overly concerned: we store a `usize` `ComponentId` anyways, and `TypeId` collisions will cause horrifying problems (almost certainly UB) for Bevy, so getting this right is important for us.

**Comment by scottmcm** [comment]
> presumably then the compiler's stable hash API is sufficiently generic over return type

Turns out it [supports exactly `u64` and `u128`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/stable_hasher/trait.StableHasherResult.html) 🙂

**Comment by james7132** [comment]
Following up on the suggestion of a crater run and @alice-i-cecile's comment on its impact to ecosystem crates, would it be possible to do a check against all 'static types in crates in the ecosystem to ensure we evaluate the real-world likelihood of a collision? To me, this is potentially impacting the performance of any code reliant on `TypeId` without an assurance that this actually addresses the issue, even in the short to medium term future while a "perfect" solution is being worked on.

**Comment by saethlin** [comment]
A collision has already been reported: https://github.com/rust-lang/rust/issues/10389#issuecomment-633328889 so I do not think the likelihood of a "real-world" collision is in doubt. The problem being worked on here can be hit and is hit. The implementation proposed here isn't even a proper fix, out of concern for performance it just makes it less likely to encounter the unsoundness.

Separately I do not think it is feasible to check against all `'static` types in the ecosystem. Crater, the existing system that might power this, just _tries_ to build every published crate. It does not always succeed. Crater does capture stdout/stderr of a build, but it has some rather severe resource limits which would result in most logs of all `TypeId`s in a program being truncated. If someone were willing to run their own hacked crater, maybe that person could get a survey of the crates that crater can successfully compile.

**Comment by Badel2** [comment]
@james7132 I was also concerned about performance in the past, but keep in mind that any performance sensitive code such as an `AnyMap` can just use the top 64 or even 32 bits of the TypeId. Also this PR is benchmarked and you see there are no noticeable regressions.

The probability of collisions has also been discussed multiple times in the past and it is high enough, although that one example of it happening in practice https://github.com/rust-lang/rust/issues/10389 is either not confirmed or the compiler was using an insecure hash function back then, can't remember the details now.

**Comment by eddyb** [comment]
AFAIK, all past collisions weren't *hash (output)* collisions but *input* collusions (equivalent to mangling ambiguities - imagine if, idk, `((), u8)` and `(u8,)` mangled the same, while still being different types).

You will never solve an *input* collusion with bigger hash *outputs*, nor with storing the whole input (i.e. the mangled type).
They're just unsoundness-causing compiler bugs, not probabilistic failure modes.
(I believe some of those bugs resulted in `Hash` being refined to more aggressively disambiguate things by prefixing/suffixing them etc.)

The only way I can imagine catching those is keeping a list of generated `TypeId`s in `rmeta` (same way we keep a list of "monomorphic instances codegen'd in this crate") and checking those lists between dependencies, while assuming the in-memory and `rmeta` representations of types is *complete* (which is needed for the soundness of rustc even ignoring `TypeId` entirely)

**Comment by nbdd0121** [comment]
> @james7132 I was also concerned about performance in the past, but keep in mind that any performance sensitive code such as an `AnyMap` can just use the top 64 or even 32 bits of the TypeId. Also this PR is benchmarked and you see there are no noticeable regressions.
> 
> The probability of collisions has also been discussed multiple times in the past and it is high enough, although that one example of it happening in practice #10389 is either not confirmed or the compiler was using an insecure hash function back then, can't remember the details now.

But `TypeId` is opaque and you can't directly get the bits.

**Comment by CAD97** [comment]
> But `TypeId` is opaque and you can't directly get the bits.

The use of `<TypeId as Hash>::hash` to extract a sufficiently hashed/mixed `u64` identifier through `Hasher::hash_u64` is an informally pseudo-blessed pattern for this. Merging this PR still doesn't officially endorse this, but it does add to the implicit acknowledgement that this is a legitimate pattern, since we're deliberately not breaking it.

It should be viable to call `hash_u64` twice, once with the high and once with the low part, to give the full state to other hashers without breaking most `TypeIdHasher`s. (For this to break the hasher, it'd have to include extra overhead in an `assert_eq!(self.state, 0, "TypeId should call hash_u64 exactly once")` or using an `Option`, and the point of such hashers is to minimize overhead.) However, I agree that given `Hash`'s fixed `u64` sized output, giving a single `u64` is sufficient for the interface.

Where we might consider giving the full state instead is for custom hashers which maintain and provide access to larger outputs. If we want to ever provide more state when hashing `TypeId` via `Hash` than just a single "sufficiently prehashed" `u64`, we should probably do it concurrently with this size change. If we want to fully commit to the ability to extract a single `u64`, it might make sense to add a method to do that specifically, as well as documenting that that's how the `Hash` impl behaves. The status quo is leaving it formally relying on implementation details but in an increasingly ossified manner that's unlikely (but still technically allowed) to change.

**Comment by thomcc** [comment]
> Merging this PR still doesn't officially endorse this, but it does add to the implicit acknowledgement that this is a legitimate pattern, since we're deliberately not breaking it

This is why I have a FIXME on this being discussed more.

> It should be viable to call `hash_u64` twice, once with the high and once with the low part, to give the full state to other hashers without breaking most `TypeIdHashers`.

This is what the initial version of the PR did, as you may notice. I changed it in [`50ae15e` (#109953)](https://github.com/rust-lang/rust/pull/109953/commits/50ae15ede9be80c090a51d6a89218a32169b50df).

> (For this to break the hasher, it'd have to include extra overhead in an `assert_eq!(self.state, 0, "TypeId should call hash_u64 exactly once")`

I've done this with `ThreadId` before FWIW (long ago, and just for improved logging), so I don't think it's hard to imagine it happening.

File: library/core/src/any.rs
Comment (WaffleLapkin) [reply library/core/src/any.rs:722]: ping: the MCP was completed, do you still want to nominate this for `T-libs`?

(IMO this impl can be changed after/asynchronously to this PR — as discussed here previously, this implementation is totally fine for the `Hash` trait)

**Comment by pnkfelix** [comment]
Discussed in T-compiler triage meeting: https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-05-11/near/357601747

Nominating for T-libs discussion since @thomcc sounded like they wanted that, see: https://github.com/rust-lang/rust/pull/109953/files#r1168527605

**Comment by Amanieu** [comment]
We discussed this in the libs meeting today. We're happy to keep the hash implementation as it is with just using the low 64 bits of the TypeId. However we would like one changes to be made to the comment:
- Compatibility with the ecosystem is *not* the reason we do this: crates should not be making assumptions on which methods of the `Hash` trait are called, and we reserve the right to change this in the future with no warning.

**Comment by thomcc** [comment]
I ended up squashing the changes as part of the rebase, to avoid having to resolve conflicts in some of the over-complicated intermediate approaches I tried. The only material change in the update (besides the rebase) is a rewritten comment inside the Hash implementation.

File: library/core/src/any.rs
Comment (bjorn3) [inline library/core/src/any.rs:716]: typo
```suggestion
        //   redundant compared to choosing (the lower) 64 bits arbitrarily.
```

**Review by WaffleLapkin — APPROVED** [review]

**Comment by WaffleLapkin** [comment]
Given the T-libs approval modulo comment in https://github.com/rust-lang/rust/pull/109953#issuecomment-1570969315 and the comment being fixed,

@bors r+ rollup

🎉

**Pull request merged**
